### PR TITLE
[Split Build] Build libtorch_python using external libraries (specifically c10)

### PR DESCRIPTION
**Pull request by PaliC**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126813
* #126826
* #126590
* #126699
* #126564
* __->__ #126325
* #126328

Builds a second wheel for torch_python which builds off the wheel created in the last part of the stack. Specifically it uses c10 from a seperate wheel

Testing: Ran the smoke tests in builder and seem to pass about as well as the normal build (passes everything outside of `test_cuda_runtime_errors_captured`

Verification 
```
sahanp@devgpu086 ~/pytorch ((6c448c29…)) [128]> ls /home/sahanp/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/libtorch/lib/                                                                                                      (pytorch-3.10) 
libbackend_with_compiler.so*  libc10_cuda.so*  libc10d_cuda_test.so*  libc10.so*  libcaffe2_nvrtc.so*  libjitbackend_test.so*  libtorchbind_test.so*  libtorch_cpu.so*  libtorch_cuda.so*  libtorch_global_deps.so*  libtorch.so*
sahanp@devgpu086 ~/pytorch ((6c448c29…))> ldd /home/sahanp/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/torch/lib/libtorch_python.so                                                                                            (pytorch-3.10) 
        linux-vdso.so.1 (0x00007ffeeb43a000)
        libshm.so => /home/sahanp/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/torch/lib/libshm.so (0x00007f512f87b000)
        libtorch.so => /home/sahanp/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/libtorch/lib/libtorch.so (0x00007f512f876000)
        libnvToolsExt.so.1 => /usr/local/cuda/lib64/libnvToolsExt.so.1 (0x00007f512e200000)
        libtorch_cpu.so => /home/sahanp/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/libtorch/lib/libtorch_cpu.so (0x00007f5122a00000)
        libtorch_cuda.so => /home/sahanp/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/libtorch/lib/libtorch_cuda.so (0x00007f5110600000)
        libcudart.so.12 => /usr/local/cuda/lib64/libcudart.so.12 (0x00007f5110200000)
        libc10_cuda.so => /home/sahanp/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/libtorch/lib/libc10_cuda.so (0x00007f512f7d9000)
        libc10.so => /home/sahanp/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/libtorch/lib/libc10.so (0x00007f512e522000)
        libcudnn.so.8 => /usr/local/cuda-12.1/targets/x86_64-linux/lib/libcudnn.so.8 (0x00007f510fe00000)
        libstdc++.so.6 => /home/sahanp/.conda/envs/pytorch-3.10/lib/libstdc++.so.6 (0x00007f510fc1d000)
        libgcc_s.so.1 => /home/sahanp/.conda/envs/pytorch-3.10/lib/libgcc_s.so.1 (0x00007f512e507000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f510fa00000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f512f886000)
        libmkl_intel_lp64.so.2 => /home/sahanp/.conda/envs/pytorch-3.10/lib/libmkl_intel_lp64.so.2 (0x00007f510e600000)
        libmkl_gnu_thread.so.2 => /home/sahanp/.conda/envs/pytorch-3.10/lib/libmkl_gnu_thread.so.2 (0x00007f510c800000)
        libmkl_core.so.2 => /home/sahanp/.conda/envs/pytorch-3.10/lib/libmkl_core.so.2 (0x00007f5108200000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f512e42a000)
        libgomp.so.1 => /home/sahanp/.conda/envs/pytorch-3.10/lib/libgomp.so.1 (0x00007f5110507000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f512e425000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f512e420000)
        librt.so.1 => /lib64/librt.so.1 (0x00007f512e41b000)
        libcupti.so.12 => /usr/local/cuda/lib64/libcupti.so.12 (0x00007f5107800000)
        libcusparse.so.12 => /usr/local/cuda/lib64/libcusparse.so.12 (0x00007f50f8200000)
        libcufft.so.11 => /usr/local/cuda/lib64/libcufft.so.11 (0x00007f50eec00000)
        libcusolver.so.11 => /usr/local/cuda/lib64/libcusolver.so.11 (0x00007f50dc600000)
        libcurand.so.10 => /usr/local/cuda/lib64/libcurand.so.10 (0x00007f50d6000000)
        libcublas.so.12 => /usr/local/cuda/lib64/libcublas.so.12 (0x00007f50cf600000)
        libcublasLt.so.12 => /usr/local/cuda/lib64/libcublasLt.so.12 (0x00007f50ae600000)
        libutil.so.1 => /lib64/libutil.so.1 (0x00007f512e412000)
        libnvJitLink.so.12 => /usr/local/cuda/lib64/libnvJitLink.so.12 (0x00007f50ab600000)
sahanp@devgpu086 ~/pytorch ((6c448c29…))> ls /home/sahanp/.conda/envs/pytorch-3.10/lib/python3.10/site-packages/torch/lib/                                                                                                               (pytorch-3.10) 
libshm.so*  libtorch_global_deps.so*  libtorch_python.so*
```

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal

### CODE REVIEW

**Review by atalman — COMMENTED** [review]
Why ideep update was done ?
Could you please post full build log somewhere ?

**Review by atalman — CHANGES_REQUESTED** [review]
Can you please specify how both pacakges are installed and run in the test plan

File: c10/CMakeLists.txt [hunk 1]
```diff
@@ -12,120 +12,127 @@ set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
...
-    c10 PUBLIC
-    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../>
-    $<BUILD_INTERFACE:${CMAKE_BINARY_DIR}>
-    $<INSTALL_INTERFACE:include>)
+  target_include_directories(
+      c10 PUBLIC
+      $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/../>
+      $<BUILD_INTERFACE:${CMAKE_BINARY_DIR}>
+      $<INSTALL_INTERFACE:include>)
+  set(C10_LIB c10)
```
Comment (atalman) [inline c10/CMakeLists.txt:134]: Please help me understand BUILD_LIBTORCHLESS -> is ON for new package ?
And NOT BUILD_LIBTORCHLESS is default path ?

However This 
```
set(C10_LIB c10)
``` 
is used when NOT BUILD_LIBTORCHLESS, is this an error ?

Comment (PaliC) [reply c10/CMakeLists.txt:134]: The set command just sets the name (similar to a variable). It let's us reduce the 
```
IF(BUILD_LIBTORCHLESS)
...
else()
...
endif()
```

where possible

**Pull request closed without merging**